src/reconcile/rec_write.c

```diff
@@ -263,7 +263,7 @@ __reconcile(WT_SESSION_IMPL *session, WT_REF *ref, WT_SALVAGE_COOKIE *salvage, u
         btree->rec_multiblock_max = r->multi_next;
 
     /* Clean up the reconciliation structure. */
-    __rec_cleanup(session, r);
+    WT_RET(__rec_cleanup(session, r));
```

Comment:
WT-8074 should've gone in a slightly different direction to support further reconciliation changes. We should have set a boolean "panic" and had an `err:` label to catch all failures beyond the panic stage. Currently this change will break the logic added in WT-8074. If you would like I can try and implement my idea on this PR and push it.

The other options is that we fix these cases. At this stage if ret == 0 then we're not panicking. So we could do:
```
if (ret == 0 && (ret = __rec_cleanup(session, r)) != 0) {
    goto panic;
} else 
     WT_TRET(__rec_cleanup(session, r));
```

Reply:
Hrm, I seem to have misread the code. I guess I believed the comment that says "any failure past this point is a panic" :-) ... not reading carefully enough is all too common when merging.

I suspect that the tidiest way to arrange it in the long term is to move all the code past where err gets set, or at least all of it that needs to be run even if the page reconciliation fails, to its own function (which can then do normal error handling) and then panic if that fails. But maybe not, I'm falling asleep and going to need to come back tomorrow.

Will mark this PR as a draft for the moment.

Reply:
oh and feel free to push here if you want; AFAIK it should let you...

Reply:
I created a branch in your fork with a suggested change which I was going to PR in your fork. But I can't push the branch to your repro can you give me permissions?

Reply:
I have tried to, with luck it might even have worked...

Reply:
But I think you can push to this branch because you have access to the original repo, and it might be easier to just do that.

Reply:
I pushed two more commits to another branch here: https://github.com/sauclovian-wt/wiredtiger/tree/wt-8476-morebits
The first (I think) fixes my changes to not be overtly broken; the second demonstrates the change I suggested above. Didn't want to put either here directly and maybe interfere with what you might be doing, but can easily pull either or both in.

Reply:
Yeah let's go with that refactor change its much cleaner. 👍

Reply:
However, there's another point. Since I was wrong about the panic paths, this change _does_ add a new trigger for the panic, and, in fact, two of them. Do we care? I feel like there probably aren't any particularly worrisome ways for cursor_reset or cursor_close to fail, but I haven't actually gone and checked. Could eliminate some of it by closing the cursor after each page instead of trying to save it...

Reply:
(Your changes are also fine with me)

Reply:
Saving the cursor does sound better than closing it after each page, I don't think that `cursor_close` or `cursor_reset` would have likely failure scenarios and can probably be safely included in our panic reasons.

Reply:
> (Your changes are also fine with me)

The refactor is definitely better though.

Reply:
I've pushed an updated version of it here, plus the obvious next set of tidyups. I think this means we should be more or less good to go (if nothing blows up in testing, and if nobody else spots anything in the rest of the changes), so I'm going to make it not a draft again.

Other thing I guess is how to arrange things so that the ultimate commit message after merging is coherent. I don't know how that works on the merge side (having not done that for a couple years, and then in an organization that didn't approve of merge-squash) so someone tell me what to do :-)

Reply:
I can merge if required, I will work with you on a reasonable commit message.